### ISSUE TITLE
Recovering connections for pool when error occurs.

### ISSUE THREAD

**Issue by boxofrox** [blackbeam/mysql_async#18]
I've got a bit of code that uses a `mysql_async::Pool` in its own thread.  A `futures::sync::mscp` is used to transfer queries in, and data extracted from `QueryResult`s out.

After updating *mysql_async* from 0.11.1 to 0.12.2, I noticed random parsing errors in my logs:

- `EOF while reading length-encoded string`
- `Invalid length-encoded integer value`

After either of these errors occurs, my "query queue" freezes on the next query while waiting for a new connection from the `Pool`.  I'm not entirely clear on what's going on here, because I'm finding high CPU usage and a backtrace on my threads shows one thread is polling *mysql_async* for data on a `TcpStream`.  I also seem to be stuck at one connection despite having an `Opts::pool_max` set to 8.

<p><details><summary>Release Backtrace (click to expand)</summary>

```
(gdb) bt
#0  0x00007f85c9509dda in recv () from /usr/lib/libpthread.so.0
#1  0x000055b42e09b2b4 in std::sys::imp::net::Socket::recv_with_flags () at /checkout/src/libstd/sys/unix/net.rs:231
#2  std::sys::imp::net::Socket::read () at /checkout/src/libstd/sys/unix/net.rs:240
#3  std::sys_common::net::TcpStream::read () at /checkout/src/libstd/sys_common/net.rs:248
#4  std::net::tcp::{{impl}}::read () at /checkout/src/libstd/net/tcp.rs:563
#5  0x000055b42e07f187 in <mio::net::tcp::TcpStream as std::io::Read>::read ()
#6  0x000055b42e014022 in <tokio_core::net::tcp::TcpStream as std::io::Read>::read ()
#7  0x000055b42dff76cd in <mysql_async::io::Stream as futures::stream::Stream>::poll ()
#8  0x000055b42dfe154c in <futures::stream::future::StreamFuture<S> as futures::future::Future>::poll ()
#9  0x000055b42dff33af in <mysql_async::connection_like::read_packet::ReadPacket<T> as futures::future::Future>::poll ()
#10 0x000055b42dfdafc2 in <futures::future::map::Map<A, F> as futures::future::Future>::poll ()
#11 0x000055b42dfc2c66 in <futures::future::and_then::AndThen<A, B, F> as futures::future::Future>::poll ()
#12 0x000055b42dfef203 in mysql_async::conn::pool::Pool::handle_futures ()
#13 0x000055b42dfefd77 in mysql_async::conn::pool::Pool::poll ()
#14 0x000055b42dfed459 in <mysql_async::conn::pool::futures::get_conn::GetConn as futures::future::Future>::poll ()
#15 0x000055b42df08623 in <futures::future::then::Then<A, B, F> as futures::future::Future>::poll ()
#16 0x000055b42debd595 in <futures::task_impl::Spawn<T>>::enter::{{closure}} ()
#17 0x000055b42deb9203 in std::sys_common::backtrace::__rust_begin_short_backtrace ()
#18 0x000055b42debca1c in std::panicking::try::do_call ()
#19 0x000055b42e0af85f in panic_unwind::__rust_maybe_catch_panic () at /checkout/src/libpanic_unwind/lib.rs:101
#20 0x000055b42dee89d7 in <F as alloc::boxed::FnBox<A>>::call_box ()
#21 0x000055b42e0a5c6c in alloc::boxed::{{impl}}::call_once<(),()> () at /checkout/src/liballoc/boxed.rs:772
#22 std::sys_common::thread::start_thread () at /checkout/src/libstd/sys_common/thread.rs:24
#23 std::sys::imp::thread::{{impl}}::new::thread_start () at /checkout/src/libstd/sys/unix/thread.rs:90
#24 0x00007f85c950008a in start_thread () from /usr/lib/libpthread.so.0
#25 0x00007f85c902124f in clone () from /usr/lib/libc.so.6
```

</details></p>

If the future error'ed, I expect the future's state machine to move on and stop polling the connection.  I've got a bit more work ahead to sort this out, and it's a bit over my head.

That work aside, I'm trying to determine what I'm expected to do when an error occurs, because my `Pool` is losing `Connection`s and the `mysql_async::errors` I get don't deliver ownership of the connection held in the [somewhat?] active query, so I can disconnect or reset it, or even establish a new connection to replace it in the pool.

<p><details><summary>Example code (click to expand)</summary>

```rust
enum Msg {
    SelectCategories(query : String, tx_chan : futures::sync::mpsc::Sender<Result<Vec<Category>>>)
}

fn runner(db_url : String, rx : futures::sync::mpsc::Receiver<Msg>) {
    let mut core = Core::new()::expect("Unable to create event loop.");
    let handle = core.handle();

    let mut opts = OptsBuilder::from_opts(&db_url);

    opts.pool_min(Some(1_usize));
    opts.pool_max(Some(8_usize));

    let pool = Pool.new(opts, &handle);

    let msg_loop = rx.and_then(|msg| match msg {
        Msg::SelectCategories(query, tx_chan) => {
            let future = pool.get_conn()
                 .and_then(move |conn| conn.query(query.as_ref()))
                 .and_then(|qres : QueryResult<_, _>| {
                     qres.map_and_drop(|row| {
                         let (id, name) : (String, Option<String>) = mysql::from_row(row);
                         // errors appear to trigger in here.
                         Category::new(id, name)
                     })
                     .map(|(conn, rows)| rows)
                 })
                 .then(|result|
                     tx_chan.send(result);
                     // Debug statements here are reached, suggesting the current task has abandoned the qres.map_and_drop above.
                     // What happens to the connection that triggered the error now?
                     Ok(())
                 });
             
             Box::new(future) as Box<Future<Item = (), Error = _>>
        }
    });

    core.run(msg_loop.for_each(|_| OK(()))).unwrap();
}
```

</details></p>

Additional notes:
- Database is a MariaDB 10.1.28 server.

**Comment by blackbeam** [comment]
Hi!

There is no way you can utilize more than one connection with this code because Stream is sequential. You should use `Stream::buffered` (purpose of `Delay` in this code is to slow things down):

<p><details><summary>Example code (click to expand)</summary>

```rust
    #[test]
    fn issue_18() {
        use lib_futures::{Sink, Stream};
        use lib_futures::sync::mpsc;
        use futures_timer::Delay;
        use QueryResult;
        use from_row;
        use errors::Result;
        use std::time::Duration;

        #[derive(Debug)]
        struct Category(String, Option<String>);

        enum Msg {
            SelectCategories{
                query: String,
                tx_chan: mpsc::Sender<Vec<Category>>,
            }
        }

        let mut lp = Core::new().unwrap();
        let pool = Pool::new(
            format!("{}?pool_min=1&pool_max=8", &**DATABASE_URL),
            &lp.handle(),
        );

        let (tx, rx) = mpsc::channel(0);

        for _ in 0..8 {
            let tx = tx.clone();
            ::std::thread::spawn(move || {
                let mut tx = tx;
                loop {
                    let (otx, orx) = mpsc::channel(0);
                    tx = tx
                        .send(Msg::SelectCategories {
                            query: "SELECT 'name', NULL".into(),
                            tx_chan: otx,
                        })
                        .wait().unwrap();
                    let (cats_opt, _) = orx.into_future().wait().unwrap();
                    if let Some(cats) = cats_opt {
                        for cat in cats {
                        }
                    }
                }
            });
        }

        let msg_loop = rx
            .map(|msg| match msg {
                Msg::SelectCategories { query, tx_chan } => {
                    pool.get_conn()
                        .and_then(move |conn| conn.query(query))
                        .then(|qres| {
                            Delay::new(Duration::from_secs(2))
                                .then(|_| {
                                    qres.unwrap()
                                        .map_and_drop(|row| {
                                            let (id, name): (String, Option<String>) = from_row(row);
                                            Category(id, name)
                                        })
                                        .map_err(|e| {
                                            ()
                                        })
                                        .map(|(conn, rows)| {
                                            tx_chan.send(rows)
                                                .map(|tx| tx.flush())
                                                .map_err(|e| {
                                                    ()
                                                })
                                                .map(|_| conn)
                                        })
                                })
                        })
                        .map(|_| {
                            ()
                        })
                }
            })
            .buffered(8);

        lp.run(msg_loop.for_each(|_| Ok(()))).unwrap();
    }
```

</details></p>

I still need to investigate more on errors. After an error connection should either return to the pool or should be dropped (for cases when it errors in the middle of a protocol and there is nothing left to do).

**Comment by boxofrox** [comment]
> There is no way you can utilize more than one connection with this code because Stream is sequential.  You should use Stream::buffered (purpose of Delay in this code is to slow things down):

Thanks for the tip on `Stream::buffered`, I didn't know of it.

> After an error connection should either return to the pool or should be dropped (for cases when it errors in the middle of a protocol and there is nothing left to do).

This sounds good.  In my situation, I guess the connection was in the middle of a result set transfer, then after the error occurred, the same connection was used for the next query but the server wasn't having that until it finished with the previous query.

As a side note, I've found the `length-encoded` errors occur more frequently (999 out of 1000 tests) for me when I'm fetching two things (one field, one expression), and the expression has a significant amount of text.  For example,

```sql
SELECT id,
    CONCAT(' '
        , id, description_short, description_long
        , GROUP_CONCAT(category SEPARATED BY ' ')
        ) as search_text
FROM products
GROUP BY sku
;
```
This query is more reliable when I drop the `id` field from the SELECT clause.

Anyway, I've had to drop *mysql_async* from my production service, and I'm using *rust-mysql-simple* which works flawlessly thus far.  I think there are a few more bugs to sort out with the async packet parsing.  When I find time, I'll investigate further, develop a SSCCE, and open an separate issue for the `length-encoded` errors.
